### PR TITLE
feat(api): Add notifications API

### PR DESCRIPTION
**Pull request by corollari**
This PR integrates the notifications API discussed in #511

Regarding the Notifications class, I haven't implemented an api like `api.neoscan.getBalance("http://www.neoscan-testnet.io/test_net/v1/", address)` because the `notifications` api is stateful (needs to keep track of the websockets connections to be able to close them later) while `neoscan` is largely not.

### CODE REVIEW

File: packages/neon-api/src/provider/notifications/class.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      this.url = settings.networks[url].extra.notifications;
+    } else {
+      this.url = url;
+    }
+	this.websocketSubscriptions = new Map<string, WebSocket[]>();
+    log.info(`Created Notifications Provider: ${this.url}`);
+  }
+
+  public subscribe(contract: string, callback: (message: NotificationMessage) => void) {
+	  const ws = new WebSocket(this.url + "?contract=" + contract);
```
Comment (snowypowers) [inline packages/neon-api/src/provider/notifications/class.ts]: new websocket for every callback? We can do better than this. For subscriptions to the same contract, I expect that we reuse the old socket.

File: packages/neon-api/src/provider/notifications/class.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  private url: string;
+
+  public get name(): string {
+    return `Notifications[${this.url}]`;
+  }
+
+  private websocketSubscriptions: Map<string, WebSocket[]>;
+
+  public constructor(url: string) {
+    if (settings.networks[url] && settings.networks[url].extra.notifications) {
```
[the remark this answers is not in the document]
Comment (corollari) [reply packages/neon-api/src/provider/notifications/class.ts]: This was taken from the code of other providers, we'd have to change that other code also in order to maintain consistency

File: packages/neon-api/src/provider/notifications/class.ts (hunk 1 above)
Comment (corollari) [reply packages/neon-api/src/provider/notifications/class.ts]: Agree, working on it.

File: packages/neon-api/src/provider/notifications/class.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (snowypowers) [reply packages/neon-api/src/provider/notifications/class.ts]: no worries. Focus on this file for now. We can migrate the rest in another PR. This is a new feature that came with Typescript 3.7.

**Comment by corollari** [comment]
> I would avoid the use of the word Provider here as there is a specific interface for it. Reusing the name would be confusing. Consider not calling it a Provider in the docs and moving it under a different folder.

Makes sense.

> Any reason why we are not adding support for the rest of the endpoints or is that coming too?

Looking at the evolution and usage that this kind of system has had on ethereum it seems that users mostly just have two use-cases:
- Listen for events in their own smart contract
- Activate a callback when a transaction is consolidated in a block

Because of this I think that the other endpoints should be integrated in different ways (directly into the API that broadcasts transactions for example) but I believe that this other part is more controversial and wanted to get the first PR simple, that's why for now I'm only integrating a single endpoint.

> Please try to separate the code and documentation. Docs can come later when we commit the code. If there are big changes, its gonna be a headache trying to reread the docs every time something changes.

Okay.

File: packages/neon-api/src/provider/notifications/class.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      this.url = settings.networks[url].extra.notifications;
+    } else {
+      this.url = url;
+    }
+	this.websocketSubscriptions = new Map<string, WebSocket[]>();
+    log.info(`Created Notifications Provider: ${this.url}`);
+  }
+
+  public subscribe(contract: string, callback: (message: NotificationMessage) => void) {
```
[the remark this answers is not in the document]
Comment (corollari) [reply packages/neon-api/src/provider/notifications/class.ts]: Personally, when designing APIs I believe the most important thing is to adhere to standards that developers are already familiar with, as that makes the API much more easier to use in my experience. With that in mind, I came up with three different models when designing this api:
1. The model on this PR: `subscribe(contract, callback), unsubscribe(contract)`
2. The standard subscription model used in DOM APIs: `subscribe(contract, callback), unsubscribe(contract, callback)`
3. The model you proposed: `a = subscribe(contract, callback); a.unsubscribe();`

While the 3rd model is the one that I find the most elegant, I believe that the other two would be more intuitive to the developers using neon-js, because:
- 2nd model: A lot of them are probably familiar with the DOM api, given that they are using a js library
- 1st model: This is the model used by [the dAPI](https://neologin.io/api/#event-methods). I didn't design it and I believe it's not the best model but in the future I expect that a lot of the devs using neon-js on the front-end will also have to deal with the dAPI, so there's some value in having consistency there.
- 1st model: I searched "pubsub javascript" on google and two of the first results that came up were tutorials that taught how to implement a pubsub system based on the 1st model

Because of this, I chose to go with the 1st model, but I have no problem changing it if it makes more sense.

[the remark this answers is not in the document]
Comment (corollari) [reply packages/neon-api/src/provider/notifications/class.ts]: > Subscribe to all contract events using null

The reason why I didn't account for this use-case is because I believe almost nobody will use it like this, but the solution you proposed (using null) doesn't pollute the api so I'm down for adding that.

**Comment by corollari** [comment]
I believe I've now addressed all the concerns except those related to api design.

File: packages/neon-api/src/notifications/class.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+  private websocketSubscriptions: Map<string | null, WebSocket[]>;
+
+  public constructor(url: string) {
+    this.url = settings.networks[url]?.extra?.notifications ?? url;
+    this.websocketSubscriptions = new Map<string | null, WebSocket[]>();
+    log.info(`Created Notifications Provider: ${this.url}`);
+  }
+
+  public subscribe(
```
[the remark this answers is not in the document]
Comment (corollari) [reply packages/neon-api/src/notifications/class.ts]: Regarding that, I'm not really sure when to write a description for a method and when not to. Looking at the current source the only pattern I can see is that functions with lots of lines are generally annotated, is there any heuristic in place for that?

File: packages/neon-api/src/notifications/class.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    contract: string | null,
+    callback: (message: NotificationMessage) => void
+  ) {
+    const ws = new WebSocket(
```
[the remark this answers is not in the document]
Comment (corollari) [reply packages/neon-api/src/notifications/class.ts]: Yeah, given that I'll probably have to re-write everything when we change the api design, I was waiting for that.

**Comment by corollari** [comment]
I see, what kind of object were you thinking to use as `reference` on the unsubscribe calls?

File: packages/neon-api/src/notifications/class.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (snowypowers) [reply packages/neon-api/src/notifications/class.ts]: generally I would annotate the public api. So it would be the constructor, subscribe and unsubscribe here. Unsubscribe might be pretty straight forward so no need but the constructor and subscribe would benefit from some description to indicate how to use. For example, indicating that you can subscribe to all by not passing an argument.

**Comment by snowypowers** [comment]
I was thinking of just an empty object at the moment. You may consider adding maybe an identifier (maybe human readable). This paves the way for us to transform this into a class with methods if we ever need to go down the route where we want the unsubscribe/pause function on the reference.

**Comment by corollari** [comment]
So something that would work like this:
```
let object = notificationsProvider.subscribe('0xabc')
notificationsProvider.unsubscribe(object)
```
?

File: packages/neon-api/src/notifications/class.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (corollari) [reply packages/neon-api/src/notifications/class.ts]: Makes sense, I thought that would be the case but when I looked at the code of the neoCli provider there was only a single function annotated, an internal one not directly exposed. But I can see that the functions there have more descriptive names so it would make sense not to annotate them.

**Comment by snowypowers** [comment]
yeah

**Comment by corollari** [comment]
I'm currently busy with exams so I'll put this PR on hiatus and come back to it in a few weeks.

**Comment by corollari** [comment]
I've changed the API to work in the way you initially proposed:
```
let subscription = notifications.subscribe(hash, console.log)
subscription.unsubscribe()
```

File: packages/neon-api/src/notifications/class.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+   * Subscribe to event notifications of a specific contract
+   * @param contract - Hash of the contract (null for all contracts) to subscribe to
+   * @param callback - Function to call when a notification is received.
+   * @return Subscription object that can be used to cancel the subscription
+   */
+  public subscribe(
+    contract: string | null,
+    callback: CallbackFunction
+  ): Subscription {
+    if (this.subscriptions.has(contract)) {
```
Comment (snowypowers) [inline packages/neon-api/src/notifications/class.ts:41]: Optional, a cleaner flow would be to first pull out the `ContractSubscriptions` object first and just do a push. 

```
const contractSubscriptions = this.subscriptions.get(contract) ?? createWebsocketForContract(contract);
contractSubscriptions.callbacks.push(callback);
```

File: packages/neon-api/src/notifications/class.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+      } else {
+        this.unsubscribeContract(contract);
+      }
+    };
+    return new Subscription(contract, unsubscribe);
+  }
+
+  /**
+   * Unsubscribe all callbacks from a specific contract
+   * @param contract - Hash of the contract (null for all contracts) to unsubscribe callbacks from
```
Comment (snowypowers) [inline packages/neon-api/src/notifications/class.ts:79]: we will have to make it clear that the null string is a special subscription that listens to all contracts and `unsubscribeContract() !== unsubscribeAll()`.

File: packages/neon-api/src/notifications/subscription.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    this.unsubscribeFunction = unsubscribeFunction;
+    this.alive = true;
+    log.info(`Created Subscription: ${this.contract}`);
+  }
+
+  /**
+   * Unsubscribe a specific function from a contract
+   */
+  public unsubscribe() {
+    if (!this.alive) {
```
Comment (snowypowers) [inline packages/neon-api/src/notifications/subscription.ts:24]: this inner variable feels redundant when we are not throwing. The savings is pretty small for the cost of having a possible wrong state.

File: packages/neon-api/src/notifications/class.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    }
+    const unsubscribe = () => {
+      if (!this.subscriptions.has(contract)) {
+        // Needed because user might have called unsubscribeAll() before
+        return;
+      }
+      const subscriptions = this.subscriptions.get(contract)!;
+      const index = subscriptions.callbacks.indexOf(callback);
+      if (index === -1) {
+        // Could happen if unsubscribeAll() followed by subscribe() are called on the same contract
```
Comment (snowypowers) [inline packages/neon-api/src/notifications/class.ts:65]: add a test for this edge case.

File: packages/neon-api/src/notifications/class.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        }
+      };
+      this.subscriptions.set(contract, {
+        websocket: ws,
+        callbacks: [callback]
+      });
+    }
+    const unsubscribe = () => {
+      if (!this.subscriptions.has(contract)) {
+        // Needed because user might have called unsubscribeAll() before
```
Comment (snowypowers) [inline packages/neon-api/src/notifications/class.ts:59]: testcase

File: packages/neon-api/src/notifications/class.ts [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+        return;
+      }
+      const subscriptions = this.subscriptions.get(contract)!;
+      const index = subscriptions.callbacks.indexOf(callback);
+      if (index === -1) {
+        // Could happen if unsubscribeAll() followed by subscribe() are called on the same contract
+        return;
+      }
+      if (subscriptions.callbacks.length > 1) {
+        subscriptions.callbacks.splice(index, 1);
```
Comment (snowypowers) [inline packages/neon-api/src/notifications/class.ts:69]: testcase

File: packages/neon-api/src/notifications/class.ts [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+      } else {
+        this.unsubscribeContract(contract);
```
Comment (snowypowers) [inline packages/neon-api/src/notifications/class.ts:71]: testcase

File: packages/neon-api/src/notifications/class.ts [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    return new Subscription(contract, unsubscribe);
+  }
+
+  /**
+   * Unsubscribe all callbacks from a specific contract
+   * @param contract - Hash of the contract (null for all contracts) to unsubscribe callbacks from
+   */
+  public unsubscribeContract(contract: string | null) {
+    if (!this.subscriptions.has(contract)) {
+      return; // Not throwing an exception in order to allow unlimited calling of this function
```
Comment (snowypowers) [inline packages/neon-api/src/notifications/class.ts:83]: testcase

File: packages/neon-api/src/notifications/class.ts [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+    }
+    this.subscriptions.get(contract)!.websocket.close();
+    this.subscriptions.delete(contract);
+  }
+
+  /**
+   * Unsubscribe all callbacks from all contracts
+   */
+  public unsubscribeAll() {
```
Comment (snowypowers) [inline packages/neon-api/src/notifications/class.ts:92]: testcase

File: packages/neon-api/__tests__/notifications/class.ts [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+    expect(result.name).toMatch(expectedUrl);
+  });
+
+  test("Network name", () => {
+    const result = new notifications("UnitTestNet");
+    expect(result.name).toMatch(UnitTestNetUrl);
+  });
+});
+
+const contract = "0x314b5aac1cdd01d10661b00886197f2194c3c89b";
```
Comment (snowypowers) [inline packages/neon-api/__tests__/notifications/class.ts:33]: does it work with only 0x-prefixed strings? This detail should be noted.

File: packages/neon-api/src/notifications/subscription.ts (hunk 8 above)
Comment (corollari) [reply packages/neon-api/src/notifications/subscription.ts:24]: What alternative do you think would work best? Throwing exceptions?

Comment (snowypowers) [reply packages/neon-api/src/notifications/subscription.ts:24]: I meant that you could just remove the `alive` check and just call `unsubscribeFunction` since it doesn't fail.

Comment (corollari) [reply packages/neon-api/src/notifications/subscription.ts:24]: I see what you are saying, but that could result in unexpected behaviour in the following case:
```
let s1 = notifications.subscribe("0x...", console.log)
let s2 = notifications.subscribe("0x...", console.log)
s1.unsubscribe()
s1.unsubscribe() //s2 gets unsubscribed
```

With that said, this is a sort of an extreme edge case so we may want to ignore it.

Comment (snowypowers) [reply packages/neon-api/src/notifications/subscription.ts:24]: ah thats interesting then the issue here is that we are storing the reference to the callback as an unique identifier. This solves the problem by ensuring each subscription can only at most remove one callback. A little roundabout but i guess it works.

**Comment by corollari** [comment]
I'm currently reworking how the class synchronises the state of the subscriptions object with the individual subscriptions by storing callbacks along a unique identifier, therefore making it much easier to check whether the subscription has been removed before.

I'd thought about this previously but decided against it because I thought that looping over an array of callbacks (the most frequent action due to the fact that it's done every time a new event is received) would be faster compared to iterating over some other data structure such as the values of a map. Later, [your comment](https://github.com/CityOfZion/neon-js/pull/514#discussion_r407116381) made me consider that I might be wrong, so I ran a [performance benchmark](https://jsben.ch/leGlO) and, contrary to my expectations, it turns out that iterating over the values of a map only results in a slowdown of <2%, a performance hit that, in my opinion, is worth the simpler implementation, which results in less edge cases and removes the need to keep two components synchronized.

File: packages/neon-api/src/notifications/class.ts (hunk 6 above)
Comment (corollari) [reply packages/neon-api/src/notifications/class.ts]: Done in 50387ee49184e0db7caf59e2995ee40c5ad183c3

File: packages/neon-api/src/notifications/class.ts (hunk 7 above)
Comment (corollari) [reply packages/neon-api/src/notifications/class.ts]: Change the comments in ca90ea9508e29003659ea4e1e1a3395b4aa90b07 to try to make that clearer, thoughts?

File: packages/neon-api/src/notifications/subscription.ts (hunk 8 above)
Comment (corollari) [reply packages/neon-api/src/notifications/subscription.ts]: I've changed how that works to make reduce the edge cases and the complexity, see https://github.com/CityOfZion/neon-js/pull/514#issuecomment-626260820

**Comment by corollari** [comment]
Just added a bunch of tests, let me know if I'm missing anything.

**Comment by corollari** [comment]
With the addition of the last 2 tests, the coverage right now is sitting at 100%.

**Comment by snowypowers** [comment]
finally done! Thanks for the contribution!
